frappe/integrations/doctype/google_drive/google_drive.py

```diff
@@ -34,118 +29,56 @@ def validate(self):
+		r = oauth_obj.refresh_access_token(
+			self.get_password(fieldname="refresh_token", raise_exception=False)
+		)
 
 		return r.get("access_token")
 
 
-@frappe.whitelist()
-def authorize_access(reauthorize=None):
+@frappe.whitelist(methods=["POST", "GET"])
+def authorize_access(reauthorize=False, code=None):
 	"""
 	If no Authorization code get it from Google and then request for Refresh Token.
 	Google Contact Name is set to flags to set_value after Authorization Code is obtained.
 	"""
 
-	google_settings = frappe.get_doc("Google Settings")
-	google_drive = frappe.get_doc("Google Drive")
-
-	redirect_uri = (
-		get_request_site_address(True)
-		+ "?cmd=frappe.integrations.doctype.google_drive.google_drive.google_callback"
+	oauth_code = (
+		frappe.db.get_value("Google Drive", "Google Drive", "authorization_code") if not code else code
 	)
+	oauth_obj = GoogleOAuth("drive")
 
-	if not google_drive.authorization_code or reauthorize:
+	if not oauth_code or reauthorize:
 		if reauthorize:
 			frappe.db.set_value("Google Drive", None, "backup_folder_id", "")
-		return get_authentication_url(client_id=google_settings.client_id, redirect_uri=redirect_uri)
-	else:
-		try:
-			data = {
-				"code": google_drive.authorization_code,
-				"client_id": google_settings.client_id,
-				"client_secret": google_settings.get_password(
-					fieldname="client_secret", raise_exception=False
-				),
-				"redirect_uri": redirect_uri,
-				"grant_type": "authorization_code",
-			}
-			r = requests.post(get_auth_url(), data=data).json()
-
-			if "refresh_token" in r:
-				frappe.db.set_value("Google Drive", google_drive.name, "refresh_token", r.get("refresh_token"))
-				frappe.db.commit()
-
-			frappe.local.response["type"] = "redirect"
-			frappe.local.response["location"] = "/app/Form/{0}".format(quote("Google Drive"))
-
-			frappe.msgprint(_("Google Drive has been configured."))
-		except Exception as e:
-			frappe.throw(e)
-
-
-def get_authentication_url(client_id, redirect_uri):
-	return {
-		"url": "https://accounts.google.com/o/oauth2/v2/auth?access_type=offline&response_type=code&prompt=consent&client_id={}&include_granted_scopes=true&scope={}&redirect_uri={}".format(
-			client_id, SCOPES, redirect_uri
+		return oauth_obj.get_authentication_url(
+			get_request_site_address(True),
+			state={
+				"method": "frappe.integrations.doctype.google_drive.google_drive.authorize_access",
+				"redirect": "/app/Form/{0}".format(quote("Google Drive")),
+			},
 		)
-	}
-
-
-@frappe.whitelist()
-def google_callback(code=None):
-	"""
-	Authorization code is sent to callback as per the API configuration
-	"""
-	frappe.db.set_value("Google Drive", None, "authorization_code", code)
-	frappe.db.commit()
 
-	authorize_access()
+	frappe.db.set_value("Google Drive", None, "authorization_code", oauth_code)
+	r = oauth_obj.authorize(oauth_code, get_request_site_address(True))
+	frappe.db.set_value("Google Drive", "Google Drive", "refresh_token", r.get("refresh_token"))
```

Comment:
Since you're allowing GET, you may wanna add a commit down here. Also, single UPDATE query possible here

Reply:
hmm we don't need GET here i suppose ..moved it to a single place :P
https://github.com/frappe/frappe/pull/16935/files#diff-110d2f31c7a2be9ad9ee6b250d273ee033248ad49532149f1c473e21929124f3R167-R183